bochs/cpu/uintr.cc

```diff
@@ -0,0 +1,344 @@
+/////////////////////////////////////////////////////////////////////////
+// $Id$
+/////////////////////////////////////////////////////////////////////////
+//
+//   Copyright (c) 2023 Stanislav Shwartsman
+//          Written by Stanislav Shwartsman [sshwarts at sourceforge net]
+//
+//  This library is free software; you can redistribute it and/or
+//  modify it under the terms of the GNU Lesser General Public
+//  License as published by the Free Software Foundation; either
+//  version 2 of the License, or (at your option) any later version.
+//
+//  This library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//  Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public
+//  License along with this library; if not, write to the Free Software
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA B 02110-1301 USA
+//
+/////////////////////////////////////////////////////////////////////////
+
+#define NEED_CPU_REG_SHORTCUTS 1
+#include "bochs.h"
+#include "cpu.h"
+#define LOG_THIS BX_CPU_THIS_PTR
+
+#if BX_SUPPORT_UINTR && BX_SUPPORT_X86_64
+
+#include "scalar_arith.h"   // for lzcntq
+
+void BX_CPU_C::deliver_UINTR()
+{
+  BX_ASSERT(BX_CPU_THIS_PTR cr4.get_UINTR() && !uintr_masked());
+
+  if (!IsCanonical(BX_CPU_THIS_PTR uintr.ui_handler)) {
+    BX_ERROR(("UINTR.UI_HANDLER is not canonical: #GP(0)"));
+    exception(BX_GP_EXCEPTION, 0);
+  }
+
+  RSP_SPECULATIVE;
+
+  Bit64u tmp_RSP = RSP;
+  if (BX_CPU_THIS_PTR uintr.stack_adjust & 0x1)
+    RSP  = BX_CPU_THIS_PTR uintr.stack_adjust;
+  else
+    RSP -= BX_CPU_THIS_PTR uintr.stack_adjust;
+  RSP &= BX_CONST64(0xfffffffffffffff0);
+
+  push_64(tmp_RSP);
+  push_64(read_eflags());
+  push_64(RIP);
+  unsigned vector = lzcntq(BX_CPU_THIS_PTR uintr.uirr); // find #most significant bit in UIRR
+  // expected to be not 0
```

Comment:
This comment looks a bit misleading in its context, i.e. what is expected to be non-zero? ``vector`` == 0 is OK after all. A check for ``uintr.uirr`` != 0 can be added to the initial BX_ASSERT if necessary.

Reply:
Fixed comment and added assert